engine/src/cubos/engine/tools/scene_editor/plugin.cpp

```diff
@@ -0,0 +1,247 @@
+    }
+}
+
+static void checkAssetEvent(cubos::core::ecs::EventReader<AssetSelectedEvent> reader, cubos::core::ecs::Commands commands, Read<Assets> assets, Write<SceneInfo> scene) {
+    for (const auto& event : reader)
+    {
+        assets->load(event.asset);
+        if(assets->type(event.asset) == typeid(Scene)) {
+            spawnScene(event.asset, commands, assets, scene);
+        }
+    }
+}
+
+static int EntityNameFilter(ImGuiInputTextCallbackData* data)
+{ 
+    if (data->EventChar == '.')  return 1;
+    return 0;
+}
+
+/// draws the entities within a scene and allows the user to add or remove entities
+static void drawSceneEntities(std::vector<std::pair<std::string, cubos::core::ecs::Entity>>& entities, Write<SceneInfo> scene, Write<cubos::engine::tools::EntitySelector> selector,
+                              cubos::core::ecs::Commands& cmds, int hierarchyDepth)
+{
+    // Add entity to current scene (needs to be root, not a sub scene)
+    if (hierarchyDepth == 0 && ImGui::Button("Add Entity"))
+    {
+        std::string entityName = scene->name + "_entity_" + std::to_string(scene->entities.size() + 1);
+        entities.emplace_back(entityName, cmds.create().entity());
+    }
+
+    // List entities (that can be removed if it's not a sub scene and selected)
+    std::vector<int> entitiesToRemove;
+    for (unsigned int i = 0; i < entities.size(); i++)
+    {
+        auto name = entities[i].first;
+        auto handle = entities[i].second;
+        ImGui::PushID(&entities[i]);
+
+        if(hierarchyDepth == 0) {
+            char *cstr = new char[64];
+            strcpy_s(cstr, 64, name.c_str());
+
+            ImGui::InputText("", cstr, 64, ImGuiInputTextFlags_CallbackCharFilter, EntityNameFilter);
+
+            std::string new_name(cstr);
+            if(name.compare(new_name)) {
+                entities[i].first = new_name;
+            }
+
+            delete[] cstr;
+        }
+        else {
+            ImGui::BulletText(name.c_str());
+        }
+
+        ImGui::Spacing();
+        if (ImGui::Button("Select"))
+        {
+            printf("entity %s was selected\n", name.c_str());
+            selector->selection = handle;
+        }
+        if (hierarchyDepth == 0)
+        {
+            ImGui::SameLine();
+            if (ImGui::Button("Remove"))
+            {
+                entitiesToRemove.push_back(i);
+                printf("removing entity %s from %s \n", name.c_str(), scene->name.c_str());
+                cmds.destroy(handle);
+            }
+        }
+
+        ImGui::PopID();
+    }
+
+    for (const auto& entity : entitiesToRemove)
+    {
+        entities.erase(entities.begin() + entity);
+    }
+}
+
+/// recursively draws the scene hierarchy and allows the user to remove scenes
+static void drawSceneHierarchy(Write<SceneInfo> scene, cubos::core::ecs::Commands& cmds, Write<cubos::engine::tools::EntitySelector> selector, int hierarchyDepth)
```

Comment:
Avoid using `Write<T>` when its not a system - just pass a `SceneInfo&` reference.
This helps people reading the code figure out what is a system and what isn't.